PLAT-85321: Add EventCacheable to cache event

ybsung opened the pull request:
### Checklist

* [X] I have read and understand the [contribution guide](http://enactjs.com/docs/developer-guide/contributing/)
* [x] A [CHANGELOG entry](http://enactjs.com/docs/developer-guide/contributing/changelogs/) is included
* [ ] At least one test case is included for this feature or bug fix
* [x] Documentation was added or is not needed
* [ ] This is an API breaking change

### Issue Resolved / Feature Added
[//]: # (Describe the issue resolved or feature added by this pull request)

Needed to cache the event when `onScrollStop` event is bubbled.

### Resolution
[//]: # (Does the code work as intended?)
[//]: # (What is the impact of this change and *why* was it made?)

 I added `EventCacheable` component and the sample using it in VirtualList for `onScrollStop`.

### Links
[//]: # (Related issues, references)

PLAT-85321

### Comments

~I'll update CHANGELOG.md and JSDocs after 1st reivew.~

ryanjduffy commented on packages/ui/EventCacheable/EventCacheable.js:
```diff
@@ -0,0 +1,33 @@
+		}
+
+		render () {
+			const eventHandler = type ? {[type]: this.eventHandler} : null;
```
Minor optimization but this code creates 4 new objects:
* This object
* clone of props: `{...this.props}`
* clone of this object: `{...eventHandler}`
* Merged props of `this.props` clone and `eventHandler` clone

An equally readable (imo) but more efficient alternative is:

```
const props = {...this.props};
if (type) props[type] = this.eventHandler;
```

ryanjduffy commented on packages/ui/EventCacheable/EventCacheable.js:
```diff
@@ -0,0 +1,33 @@
+import React from 'react';
+import hoc from '@enact/core/hoc';
+
+const EventCacheable = hoc((config = {defaultEvent: null, type: null}, Wrapped) => {
```
This use of default parameters doesn't take advantage of the config merging the `hoc` provides. It doesn't make much difference in this case since there are no default values but, for consistency, I'd suggest using the following form which allows the author to pass partial configs which `hoc` will populate with defaults.

```
const defaultConfig = {...};
hoc(defaultConfig, (config, Wrapped) => { ... });
```

ryanjduffy commented on packages/ui/EventCacheable/EventCacheable.js:
```diff
@@ -0,0 +1,33 @@
+			this.cachedEvent = event;
+		}
+
+		eventHandler = (ev) => {
```
You'd probably want to forward this on rather than capture the event.

```js
import {forward, handle, call} from '@enact/core/handle'
eventHandler = handle(
  forward(type), // forward to consumers
  call('setEvent'), // invoke this.setEvent(ev)
).bindAs(this, 'eventHandler')
```

ybsung commented:
I reopened this PR because of following up the @ryanjduffy's comments. I think that this PR was closed by mistake. If not, please correct me.

ybsung commented:
Hello @ryanjduffy 

Thank you so much for your review. I had to concrete my PR more before requesting review. I updated the PR as you mentioned and added CHANGELOG.md & JSDocs. Could you review them?

ryanjduffy commented:
Thanks, @ybsung, I should have been more clear in my original review. I closed it because this feature isn't a pattern I think we should support within the framework. The code seems sound and likely suitable to be used within applications that would want access to an imperative API.

Glad you created the PR to spur this discussion and because it can be pointed to as an example for interested apps.
